### PR TITLE
[ONNX] Add BFloat16 type support when export to ONNX

### PR DESCRIPTION
**Pull request by hwangdeyu**
- PyTorch and ONNX has supported BFloat16, add this to unblock some mixed-precision training model.
- Support PyTorch TNLG model to use BFloat16 tensors for the inputs/outputs of the layers that run on the NPU.

### CODE REVIEW

**Review by shubhambhokare1 — COMMENTED** [review]
Overall LGTM, minor comments

**Review by shubhambhokare1 — APPROVED** [review]
LGTM, thanks!

**Review by garymm — CHANGES_REQUESTED** [review]

**Review by garymm — CHANGES_REQUESTED** [review]
Please add a unit test to catch the `scalar_type_to_pytorch_type` bug you introduced earlier.

**Review by garymm — APPROVED** [review]
Thanks. Please address comments before merging.

**Comment by hwangdeyu** [comment]
> > And I don't have the authority to merge this PR into master?
> 
> Oh that's because you've opened it against master rather than onnx_ms_1. You'll have to work with someone at FB to get it merged.
> 
> Looks like you have test failures. Please fix.
> 
> ```
> 2021-10-29T09:01:35.1870005Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset7::test_one_cast_pytorch_to_onnx_bfloat16
> 2021-10-29T09:01:35.1870932Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset7::test_type_as_bfloat16
> 2021-10-29T09:01:35.1871839Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset8::test_one_cast_pytorch_to_onnx_bfloat16
> 2021-10-29T09:01:35.1872763Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset8::test_type_as_bfloat16
> 2021-10-29T09:01:35.1873667Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_one_cast_pytorch_to_onnx_bfloat16
> 2021-10-29T09:01:35.1874536Z FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime::test_type_as_bfloat16
> ```


In the latest update, I removed all the exporter tests. 
Cause ORT only supports `bfloat16` type in gpu with the cuda condition, but we could't guarantee this condition in our tests.
```
CUDA_VERSION >= 11000 && (__CUDA_ARCH__ >= 800 || !defined(__CUDA_ARCH__))
```
https://github.com/microsoft/onnxruntime/blob/4df356d1c9f1387897ccb9df1d170ba80e929756/onnxruntime/core/providers/cuda/cuda_utils.cu#L74

I also have tested the tests in `ITP A100` machine, and all the tests are passed. So this PR is effective.

**Comment by garymm** [comment]
We should try hard to have automated test coverage for this.

I think at least some ORT ops support BFloat16 on CPU:
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/math/sign.cc#L90
https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cpu/tensor/cast_op.cc#L22

So I hope we can find at least one test that runs on ORT CPU.

If we really can't make ORT on CPU work, then it seems like we should be able to get it to run on CUDA by checking `torch.cuda.is_bf16_supported()`

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -6452,6 +6452,19 @@ def forward(self, input, other):
         model = MyModule()
         self.run_test(model, (x, y))
 
+    # ONNX supported bfloat16 for opsets >= 13
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:6455]: supported -> supports

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py [hunk 2]
```diff
@@ -85,6 +85,20 @@ def forward(self, x):
         model = amp.initialize(LinearModel(), opt_level="O2")
         self.run_test(model, input)
 
+    # ONNX supported bfloat16 for opsets >= 13
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:88]: supported -> supports

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py [hunk 3]
```diff
@@ -85,6 +85,20 @@ def forward(self, x):
...
+    @skipIfUnsupportedMinOpsetVersion(13)
+    @skipIfNoBFloat16Cuda
+    @disableScriptTest()
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:91]: add a comment about why script tests doesn't work

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py [hunk 4]
```diff
@@ -85,6 +85,20 @@ def forward(self, x):
...
+    def test_linear_bfp16(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:92]: I'm confused why this is called `linear`?

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py [hunk 5]
```diff
@@ -85,6 +85,20 @@ def forward(self, x):
...
+        class MyModule(torch.nn.Module):
+            def forward(self, x):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:94]: please add a comment explaining what in this model can't be tested on CPU.
If everything here can be tested on CPU, then just remove the test.

**Review by garymm — CHANGES_REQUESTED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 3 above)
Comment (hwangdeyu) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py]: Removed.

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 4 above)
Comment (hwangdeyu) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py]: That's my `A100` ITP local linear test name. Fixed it.

File: test/onnx/test_pytorch_onnx_onnxruntime_cuda.py (hunk 5 above)
Comment (hwangdeyu) [reply test/onnx/test_pytorch_onnx_onnxruntime_cuda.py:94]: Added the comment.

**Review by garymm — APPROVED** [review]
Thanks!

**Comment by hwangdeyu** [comment]
@malfet @msaroufim This is a PR for supporting `Bfloat16` type export to ONNX. 
Anything else required to merge this PR?

**Review by msaroufim — APPROVED** [review]

**Comment by hwangdeyu** [comment]
Thanks @msaroufim, 
Could you also help to merge this PR? Cause some common models really need the `Bfloat16` type support.

**Comment by hwangdeyu** [comment]
Hi @malfet @msaroufim, can we merge this PR？Thanks!

**Review by malfet — APPROVED** [review]

**Comment by garymm** [comment]
@malfet can you please merge?

**Comment by malfet** [comment]
@garymm this PR causes regression that looks legit, but didn't have time to investigate yet.

**Pull request closed without merging**